src/bun.js/test/snapshot.zig

```diff
@@ -0,0 +1,292 @@
+
+        const test_name_string = ZigString.init(snapshot_key);
+        const snapshot_data = this.snapshot_data catch {
+            this.update_snapshot = true;
+            return SnapshotError.SnapshotNotFound;
+        };
+        const value = snapshot_data.getIfPropertyExistsImpl(globalObject, test_name_string.ptr, @truncate(u32, test_name_string.len));
+        if (@enumToInt(value) == 0) {
+            return SnapshotError.CaseNotFound;
+        }
+        return value;
+    }
+
+    pub fn exists(this: *SnapshotFile) bool {
+        return (fs.FileSystem.FilenameStore.instance.exists(this.path.text) or
+            fs.FileSystem.DirnameStore.instance.exists(this.path.text)) and
+            (fs.FileSystem.FilenameStore.instance.exists(this.path.pretty) or
+            fs.FileSystem.DirnameStore.instance.exists(this.path.pretty));
+    }
+
+    fn compare(this: *SnapshotFile, actual: JSC.JSValue, expected: JSC.JSValue) bool {
+        const globalObject = this.globalObject orelse unreachable;
+
+        var fmt = JSC.ZigConsoleClient.Formatter{ .globalThis = globalObject };
+
+        const actual_formatted = std.fmt.allocPrint(this.allocator, "{any}", .{actual.toFmt(globalObject, &fmt)}) catch unreachable;
+        const expected_formatted = std.fmt.allocPrint(this.allocator, "{any}", .{expected.toFmt(globalObject, &fmt)}) catch unreachable;
+
+        return std.mem.eql(u8, actual_formatted, expected_formatted);
+    }
+
+    pub fn match(this: *SnapshotFile, snapshot_name: bun.string, actual: JSC.JSValue, not: bool, globalObject: *JSC.JSGlobalObject) !bool {
+        this.globalObject = globalObject;
+        var count: u32 = this.incrementCounter(snapshot_name) catch 1;
+
+        try this.tests.append(this.allocator, TestData{ .k = ZigString.init(snapshot_name), .v = actual, .count = count });
+
+        const expected = this.getSnapshotValue(snapshot_name, count, globalObject) catch |err| {
+            if (err == SnapshotError.SnapshotNotFound) {
+                Output.prettyln("    <green>Missing Snapshot file, creating one<r>\n", .{});
+                Output.flush();
+                // If the snapshot file does not exist, then we return true
+                // and update the snapshot file
+                this.update_snapshot = true;
+                return true;
+            }
+            if (this.update_snapshot or SnapshotFile.update_all_snapshots) {
+                Output.prettyln("    <green>Updating Snapshot<r>\n", .{});
+                Output.flush();
+                return true;
+            }
+            globalObject.throw("The snapshot `{s} {}` was not found in {s}", .{ snapshot_name, count, this.path.text });
+            return false;
+        };
+
+        var pass = this.compare(actual, expected);
+
+        if (not) pass = !pass;
+        if (pass) return true;
+
+        var fmt = JSC.ZigConsoleClient.Formatter{ .globalThis = globalObject };
```

Comment:
I'm not sure that we can use the console.log formatter in here. Is that what Jest does? The console.log formatter doesn't deeply visit every nested object and it doesn't visit every JSX child element